### DISCUSSION TITLE
Middleware does not run in production

### DISCUSSION THREAD

**Discussion by VogelVlug** [vercel/next.js#62283] [Help] [answered]
### Summary

I've created a very minimal app (for a teleprompter tool) at https://github.com/Wursteintopf/teleprompter. For some reason my middleware (placed at src/middleware.ts as it should be) runs completely fine in development, but it does not get triggered on production. So when starting next with `next dev` it works as it should but with `next build` and `next start` is does not get triggered. I have absolutely no idea why this is happening since this is not my first project with next and I've never experienced this issue before. Is this a bug in the current release? Or am I missing something obvious?

### Example

https://github.com/Wursteintopf/teleprompter

**Comment by VogelVlug** [comment]
Anyone? 🥲

**Reply by mubarakshow** [reply]
This worked for me.
https://stackoverflow.com/a/75242787/8599881

**Comment by icyJoseph** [comment]
Rather weird... I found an issue in the supabase/ssr Github, but it doesn't sound like it is the same issue at all..

Anyway, since its been two days without a solution, could you try to empty the contents of the middleware and just return `NextResponse.next()`? 

```js
export function middleware() {
   console.count("Middleware");
   return NextResponse.next();
}
```

Does that run?

---

Here is the issue I found: https://github.com/supabase/auth-helpers/issues/673

**Comment by VogelVlug** [comment]
Rather weird indeed 😕

But no, even this minimal middleware does not run in production mode:

```
import { NextResponse } from "next/server";

export async function middleware() {
  console.count("Middleware");
  return NextResponse.redirect(new URL("https://google.de"));
}

export const config = {
  // Matcher ignoring `/_next/` and `/api/`
  matcher: ["/((?!api|auth|_next/static|_next/image|favicon.ico).*)"],
};
```

I even tried a redirect in case it just silences the console for some reason, but it does not run. Again runs flawlessly when starting with `next dev` though.

**Reply by icyJoseph** [reply]
And it is failing "locally" too right? I mean a local production build. 

I have heard before of people having issues with GetServerSideProps, when GetStaticProps was used, but this is really strange to be fair. 

Does it help to change the Next version, or install `next@canary`?

**Reply by VogelVlug** [reply]
Yes, it is failing "locally" too. Both on my server aswell as on my local machine it works with `next dev` but not with a production build.

So far I've tried 14.1.0 (so the latest) and 14.0.4 which I have running in another project where the middleware works as expected. Due to your suggestion I just tried 14.1.1-canary.69, but fails there aswell.

**Reply by VogelVlug** [reply]
Ahaaaaa, wait a second. I think I am slowly getting to the bottom of this, it seems to be an issue with my docker setup. I've just ran `next build` and `next start` directly via yarn on my machine instead of in the docker container and then it works. Maybe I am forgetting to correctly mount some folder or files that are needed for the production build.

**Reply by VogelVlug** [reply, marked as answer]
So, a restructure of the docker setup indeed did the trick. To be honest, I am still not entirely sure which file I was missing to mount correctly, but it works now. Thanks for pointing me in the right direction @icyJoseph!

**Reply by simovicaleksa** [reply]
Can you please send your new Dockerfile? 🙏 I am using the one next.js has officially provided, and I am still having this issue with my middleware in production...

**Reply by VogelVlug** [reply]
@Aleksa1312 Sure, you can have a look at my code in my repo that I also linked above, this is still the final version I am using for this project: https://github.com/Wursteintopf/teleprompter

**Reply by mubarakshow** [reply]
This worked for me
https://stackoverflow.com/a/75242787/8599881

**Comment by MasterProvider** [comment]
also failed fetch in production mode
```
//middleware.js
const user = await fetch(process.env.API+'sample/data',
    { 
      method: 'POST',
      cache: 'no-store',
      headers:{
          'Content-Type': 'application/json',
      },
      body:JSON.stringify({token:token}),
    }
    ).then((res)=>res.json())
     .catch((error)=> new Error(error));
    console.log('middleware:',user);
```

**Reply by yteruel31** [reply]
@MasterProvider did you found a solution about this? I've the same issue

```
[cause]: Error [ConnectTimeoutError]: Connect Timeout Error (attempted address: mybackend.com:443, timeout: 60000ms)
      at d (.next/server/middleware.js:27:37849)
      at Immediate._onImmediate (.next/server/middleware.js:27:37477) {
    code: 'UND_ERR_CONNECT_TIMEOUT'
}
```
I'm on `15.2.1-canary.2` and using `nodeMiddleware`
